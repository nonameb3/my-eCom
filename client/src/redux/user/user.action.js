export default function(user) {
  return  {
    type: "SET_CURRENT_USER",
    payload: user
  }
}