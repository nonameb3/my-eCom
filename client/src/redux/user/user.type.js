export const SET_CURRENT_USER = 'SET_CURRENT_USER';
export const GOOGLE_SIGN_IN_START = 'GOOGLE_SIGN_IN_START';
export const EMAIL_SIGN_IN_START = 'EMAIL_SIGN_IN_START';
export const SIGN_IN_SUCCESS = 'SIGN_IN_SUCCESS';
export const SIGN_IN_FAILURE = 'SIGN_IN_FAILURE';
