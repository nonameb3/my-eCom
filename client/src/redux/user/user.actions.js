import * as TYPE from "./user.type";

export const setCurrentUser = user => ({
  type: TYPE.SET_CURRENT_USER,
  payload: user
});

export const googleSignInStart = () => ({
  type: TYPE.GOOGLE_SIGN_IN_START
});

export const emailSignInStart = input => ({
  type: TYPE.EMAIL_SIGN_IN_START,
  payload: input
});

export const signInSuccess = user => ({
  type: TYPE.SIGN_IN_SUCCESS,
  payload: user
});

export const signInFailure = error => ({
  type: TYPE.SIGN_IN_FAILURE,
  payload: error
});

export const signOutStart = () => ({
  type: TYPE.SIGN_OUT_START
});

export const signOutSuccess = () => ({
  type: TYPE.SIGN_OUT_SUCCESS
});

export const signOutFailure = error => ({
  type: TYPE.SIGN_OUT_FAILURE,
  payload: error
});

export const checkUserAuthentication = () => ({
  type: TYPE.CHECK_USER_SESSION
});

export const signUpStart = formInput => ({
  type: TYPE.SIGN_UP_START,
  payload: formInput
})

export const signUpSuccess = user => ({
  type: TYPE.SIGN_UP_SUCCESS,
  payload: user
})

export const SignUpFailure = error => ({
  type: TYPE.SIGN_UP_FAILURE,
  payload: error
})
