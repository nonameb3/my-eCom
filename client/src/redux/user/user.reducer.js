import * as TYPE from "./user.type";

const INITIAL_STATE = {
  currentUser: null,
  error: null
};

export default function userReducer(state = INITIAL_STATE, action) {
  switch (action.type) {
    case TYPE.SIGN_IN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        error: null
      };
    case TYPE.SIGN_OUT_SUCCESS:
      return {
        ...state,
        currentUser: null,
        error: null
      }
    case TYPE.SIGN_IN_FAILURE:
    case TYPE.SIGN_OUT_FAILURE:
    case TYPE.SIGN_UP_FAILURE:
      return {
        ...state,
        error: action.payload
      };
    default:
      return state;
  }
}
