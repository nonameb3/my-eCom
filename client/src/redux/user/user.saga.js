import { takeLatest, put, all, call } from "redux-saga/effects";

import {
  googleProvider,
  auth,
  createUserProfileDocument,
  getCurrentUser
} from "../../firebase/firebase.utill";
import {
  signInSuccess,
  signInFailure,
  signOutSuccess,
  signOutFailure,
  signUpSuccess,
  SignUpFailure
} from "./user.actions";
import * as TYPE from "./user.type";

export function* processUserSnapshot(user, additionalData) {
  try {
    const userRef = yield call(createUserProfileDocument, user, additionalData);
    const userSnapshot = yield userRef.get();

    // put is send data back to redux
    yield put(
      signInSuccess({
        id: userSnapshot.id,
        ...userSnapshot.data()
      })
    );
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* siginWithGoogle() {
  try {
    const { user } = yield auth.signInWithPopup(googleProvider);
    yield processUserSnapshot(user);
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* signinWithEmail({ payload: { email, password } }) {
  try {
    const { user } = yield auth.signInWithEmailAndPassword(
      email.trim(),
      password
    );
    yield processUserSnapshot(user);
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* isUserAuthenticated() {
  try {
    const userAuth = yield getCurrentUser();
    if (!userAuth) return;
    yield processUserSnapshot(userAuth);
  } catch (error) {
    yield put(signInFailure(error));
  }
}

export function* signOut() {
  try {
    yield auth.signOut();
    yield put(signOutSuccess());
  } catch (error) {
    yield put(signOutFailure(error));
  }
}

export function* signUpStart(props) {
  const { email, password } = props.payload.additionalData;
  try {
    const { user } = yield auth.createUserWithEmailAndPassword(
      email.trim(),
      password
    );

    yield put(signUpSuccess(user));
    yield processUserSnapshot(user, props.payload.additionalData);
    props.payload.history.push("/");
  } catch (error) {
    yield put(SignUpFailure(error))
  }
};

// handle function
export function* onGoogleSignInStartSaga() {
  yield takeLatest(TYPE.GOOGLE_SIGN_IN_START, siginWithGoogle);
}

export function* onEmailSignInStartSaga() {
  yield takeLatest(TYPE.EMAIL_SIGN_IN_START, signinWithEmail);
}

export function* onCheckUserSaga() {
  yield takeLatest(TYPE.CHECK_USER_SESSION, isUserAuthenticated);
}

export function* onSignOut() {
  yield takeLatest(TYPE.SIGN_OUT_START, signOut);
}

export function* onSignUpStart() {
  yield takeLatest(TYPE.SIGN_UP_START, signUpStart)
}

export function* userSagas() {
  yield all([
    call(onGoogleSignInStartSaga),
    call(onEmailSignInStartSaga),
    call(onCheckUserSaga),
    call(onSignOut),
    call(onSignUpStart),
  ]);
}
