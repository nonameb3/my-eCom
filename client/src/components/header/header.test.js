import React from "react";
import { shallow } from "enzyme";

import Header from "./header.component";
import CartDropdown from "../cart-dropdown/cart-dropdown.component";
import { signOutStart } from "../../redux/user/user.actions";

let wrapper;
const mockDispatch = jest.fn();
const mockState = {
  currentUser: { id: 1, name: "test" },
  hidden: false
};
jest.mock("react-redux", () => ({
  useSelector: jest.fn(() => mockState),
  useDispatch: () => mockDispatch
}));

describe("header component", () => {
  beforeEach(() => {
    wrapper = shallow(<Header />);
  });

  it("should render header component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render cart-dropdown component", () => {
    expect(wrapper.find(CartDropdown).exists()).toBe(true);
  });

  it("should action logout when click sign-out", () => {
    wrapper.find("#signout-btn").simulate("click");
    expect(mockDispatch).toHaveBeenCalledWith(signOutStart());
  });
});
