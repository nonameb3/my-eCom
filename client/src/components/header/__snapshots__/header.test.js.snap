// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header component should render header component 1`] = `ShallowWrapper {}`;
